ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java

```diff
@@ -359,6 +365,9 @@ protected void validateLogEntry(LogEntryProto entry) {
               "is greater than 1, entry: %s",
           lastTermIndex.getIndex(), latestSnapshotIndex, entry);
     } else {
+      // No logs are present. Check state machine's snapshot index.
+      updateSnapshotIndex(snapshotIndexSupplier.getAsLong());
+      latestSnapshotIndex = getSnapshotIndex();
```

Comment:
Updating in validateLogEntry is expensive since it is called for every entry.  How about updating it in purge?
```
+++ b/ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java
@@ -357,6 +357,7 @@ public class SegmentedRaftLog extends RaftLog {
   protected CompletableFuture<Long> purgeImpl(long index) {
     try (AutoCloseableLock writeLock = writeLock()) {
       SegmentedRaftLogCache.TruncationSegments ts = cache.purge(index);
+      updateSnapshotIndexFromStateMachine();
       LOG.debug("purging segments:{}", ts);
       if (ts != null) {
         Task task = fileLogWorker.purge(ts);
```
where updateSnapshotIndexFromStateMachine() is a new method in RaftLog.java
```
+++ b/ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java
@@ -153,6 +153,10 @@ public abstract class RaftLog implements RaftLogSequentialOps, Closeable {
     return false;
   }
 
+  protected void updateSnapshotIndexFromStateMachine() {
+    updateSnapshotIndex(getSnapshotIndexFromStateMachine.getAsLong());
+  }
+
```

Reply:
A good point. But we don't expect it to be called for every validate call since it is in updated only in the case when there are no logs. That can happen only once after every purge. But, I agree we can move this to purge.